### PR TITLE
JAVA-151: Make it possible to register for SchemaChange Events.

### PR DESCRIPTION
**Pull request by adutra**
This commit contains a substantial amount of code originally contributed by Robert Stupp (@snazy).

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 1]
```diff
@@ -2177,15 +2208,34 @@ public void runMayThrow() throws InterruptedException, ExecutionException {
                                 submitSchemaRefresh(scc.keyspace, scc.table);
                             break;
                         case DROPPED:
-                            if (scc.table.isEmpty())
-                                manager.metadata.removeKeyspace(scc.keyspace);
-                            else {
+                            if (scc.table.isEmpty()) {
+                                final KeyspaceMetadata removed = manager.metadata.removeKeyspace(scc.keyspace);
+                                if(removed != null) {
+                                    executor.submit(new Runnable() {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java:2232]: Broadcasting of schema change events is done in the manager's executor so that it doesn't block the I/O thread that received the event notification.

File: driver-core/src/main/java/com/datastax/driver/core/DefaultResultSetFuture.java [hunk 2]
```diff
@@ -82,14 +82,34 @@ public void onSet(Connection connection, Message.Response response, ExecutionInf
                                         // We'll add it back if CASSANDRA-5358 changes that behavior
                                         //if (scc.keyspace.equals(session.poolsState.keyspace))
                                         //    session.poolsState.setKeyspace(null);
-                                        session.cluster.manager.metadata.removeKeyspace(scc.keyspace);
+                                        final KeyspaceMetadata removed = session.cluster.manager.metadata.removeKeyspace(scc.keyspace);
+                                        if(removed != null) {
+                                            session.cluster.manager.executor.submit(new Runnable() {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DefaultResultSetFuture.java:87]: Same here: broadcasting of schema change events si done in the manager's executor so that it doesn't block the I/O thread that received the event notification.

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java [hunk 3]
```diff
@@ -94,27 +94,43 @@ synchronized void rebuildSchema(String keyspace, String table, ResultSet ks, Res
...
 
-            // If keyspace is null, it means we're rebuilding from scratch, so
-            // remove anything that was not just added as it means it's a dropped keyspace
-            if (keyspace == null) {
-                Iterator<String> iter = keyspaces.keySet().iterator();
-                while (iter.hasNext()) {
-                    if (!addedKs.contains(iter.next()))
-                        iter.remove();
+            Set<String> removedKs = new HashSet<String>();
+            for (String ksName : keyspaces.keySet()) {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: Handling dropped keyspaces is tricky: I prefer to call `removeKeyspace` to take care of token maps, but we also need to broadcast dropped events for all tables in the dropped keyspace as well. On the other hand, `removeKeyspace` is also called from outside this method, so it should not directly trigger broadcasting of schema change events.

File: driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java [hunk 4]
```diff
@@ -68,81 +99,132 @@ public void teardown() {
...
-
+    public void should_notify_of_table_creation(String keyspace) throws InterruptedException {
+        execute(CREATE_TABLE, keyspace);
         for (Metadata m : metadatas)
-            assertThat(m.getKeyspace(keyspace).getTable("table1").getColumn("j"))
-                .isNotNull();
+            assertThat(m.getKeyspace(keyspace).getTable("table1")).isNotNull();
+        for (SchemaChangeTestListener listener : listeners) {
+            ArgumentCaptor<TableMetadata> added = ArgumentCaptor.forClass(TableMetadata.class);
+            verify(listener, atLeastOnce()).onTableAdded(added.capture());
```
Comment (adutra) [inline driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java]: This should be `times(1)` but for a strange reason, when running with Maven this yields unexpected numbers (running with IntelliJ is fine though).

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java [hunk 5]
```diff
@@ -150,13 +170,73 @@ private void buildTableMetadata(KeyspaceMetadata ksm, List<Row> cfRows, Map<Stri
                         cols = Collections.<String, ColumnMetadata.Raw>emptyMap();
                     }
                 }
-                TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+
+                TableMetadata tm = TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+                TableMetadata tmPrevious = ksm.tables.get(tm.getName());
+                if (tmPrevious == null) {
+                    ksm.add(tm);
+                    triggerOnTableAdded(tm);
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: There is one caller of this code (refreshSchema via ControlConnection#tryConnect) which can be called in a separate executor than the main cluster executor (either the main thread or in reconnectionExecutor) are we worried about schema listeners blocking those threads?

Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: That's true; I left it that way because tryConnect() already blocks a lot (e.g. reading system tables) - even JAVA-657 won't make it any better (it will debounce all other invocations, but not the one from tryConnect). Do you think it would be wiser to call triggerOnXX() on the main executor? I don't like much the idea of having client code executed on the driver's main executor.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: Ah, do you mean the trigger methods in general would always be submitted on `cluster.manager.executor` inside their method definition?   That makes sense to me.  Although, how would it work when something executing inside of that executor calls these methods?  Not sure if we handle that elsewhere, but was just curious.

Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: It already happens with JAVA-657 (admittedly in rare situations: maybeRefreshSchemaAndSignal() for example). But on a second thought, I think that having client code executed in that executor could be a bad idea, so I'd be in favor of leaving this the way it is now.

**Comment by olim7t** [comment]
Since `SchemaChangeListener` is a new type, it would be nice to have `onRegister` and `onUnregister`, like we added for other listeners on the 2.2 branch. I'm going to add it.

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 3 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: > we also need to broadcast dropped events for all tables in the dropped keyspace as well

Do we? I think it's implicit that if a keyspace got dropped, everything in it got dropped.

Apart from that, couldn't we do without the `removedKs` set?

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 5 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: I agree to leave it the way it is. If listeners block for an unreasonable amount of time, it's going to block _some_ driver thread anyway. I'll amend the doc to mention that listeners should not block.

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 3 above)
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: You are right, broadcasting these events could be overwhelming in cases where the dropped keyspace contains several tables. I will amend that.
As for the `removedKs` set, I created it because we are iterating over `keypsaces.keySet()` and, even if the actual implementation accepts concurrent modifications, the general contract of that method says that "If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined".

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 5 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: :+1: .  In that case we should get rid of the code in Cluster#handle when handling a SCHEMA_CHANGE that runs it in cluster.manager.executor?

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 3 above)
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: Another killer argument in favor of not broadcasting events for tables in a dropped keyspace:  it seems that Cassandra already does exactly that :) : it first sends a DROPPED TABLE event for each table, then a DROPPED KEYSPACE. In the end, the user will get notifications for each table. That also explained why my tests were seeing duplicate events.

File: driver-core/src/main/java/com/datastax/driver/core/Metadata.java (hunk 5 above)
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Metadata.java]: Sorry I wasn't clear at all in my last comment: currently _all_ schema change events are processed in the main executor _except_ in tryConnect(). What I meant is that I was in favor of leaving it that way, _even knowing that it's potentially a bad idea to let user code go in that executor_, because like Olivier said it will fatally block some of the driver's threads: e.g. the handle() method is executed by an IO thread, and blocking that one could be even worse.

File: driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java [hunk 6]
```diff
@@ -39,24 +61,35 @@
...
+        cluster1 = Cluster.builder().addContactPoint(CCMBridge.ipOfNode(1)).build();
         cluster2 = Cluster.builder().addContactPoint(CCMBridge.ipOfNode(1)).build();
 
-        metadatas = Lists.newArrayList(cluster.getMetadata(), cluster2.getMetadata());
+        metadatas = Lists.newArrayList(cluster1.getMetadata(), cluster2.getMetadata());
+
+        session = cluster1.connect();
+
+        cluster1.register(listener1 = spy(new SchemaChangeTestListener()));
+        cluster1.register(listener2 = spy(new SchemaChangeTestListener()));
```
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java]: Combining a spy and a custom implementation is a bit convoluted.
I understand that the custom code provides a way to wait for the async event, but we could use a pure mock and [timeout()](http://docs.mockito.googlecode.com/hg/org/mockito/verification/VerificationWithTimeout.html) for the same purpose.

Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/SchemaChangesTest.java]: I'm going to try it and I'll push the change if it works.

**Comment by tolbertam** [comment]
Added a test and a small change to fix an issue I was running into.  Looking good to me :+1:
